src/pnpq/devices/waveplate_stub.py

```diff
@@ -0,0 +1,235 @@
+import logging
+import time
+
+from pnpq.errors import (
+    DevicePortNotFoundError,
+    DeviceDisconnectedError,
+    WavePlateMoveNotCompleted,
+    WavePlateHomedNotCompleted,
+    WavePlateGetPosNotCompleted,
+    WavePlateCustomRotateError,
+    WaveplateInvalidStepsError,
+    WaveplateInvalidDegreeError,
+    WaveplateEnableChannelError,
+    WaveplateInvalidMotorChannelError,
+)
+
+class Waveplate:
+    # conn: Serial
+    # serial_number: str
+    # port: str | None
+    resolution: int
+    max_steps: int
+    relative_home: float
+
+    def __init__(self):
+        # Stub Serial Number
+        # TODO: Custom serial number from initalizer
+        self.device_sn = "stubwaveplate"
+
+        self.resolution = 136533
+        self.max_steps = 136533
+        self.rotate_timeout = 10
+        self.home_timeout = 20
+        self.max_channel = 1
+        self.auto_update = False
+        # self.hub_connected = check_usb_hub_connected()
+
+        self.logger = logging.getLogger(f"{self}")
+
+        # Stub device parameters
+        self.current_position = 0
+
+        # Is connected to the device
+        self.connected = False
+
+        # Enabled channels (enable 1 by default)
+        self.enabled_channels = [1]
+
+    def __ensure_port_open(self) -> None:
+        if not self.connected:
+            self.logger.warning("Device not connected")
+            raise DeviceDisconnectedError(f"{self} is disconnected")
+
+    def __ensure_less_than_max_steps(self, steps: int) -> None:
+        if steps > self.max_steps:
+            raise WaveplateInvalidStepsError(f"Given steps: {steps} exceeds the maximum steps: {self.max_steps}")
+
+    def __ensure_valid_degree(self, degree: float) -> None:
+        if 0 <= degree <= 360:
+            return
+        raise WaveplateInvalidDegreeError(f"Invalid degree: {degree}. Degree must be in a range [0,360]")
+
+    # Maybe make this into a function wrapper to remove some redundant code?
+    def __stub_check_channel(self, chanid: int):
+        return chanid in self.enabled_channels
+
+    def connect(self):
+        self.connected = True
+        self.logger.info("Stub Waveplate Connected")
+
+    def identify(self):
+        self.__ensure_port_open()
+        # Flashes LED on real device, for stub, we will do nothing
+        self.logger.info("Stub Waveplate Identify")
+
+    def home(self):
+        self.__ensure_port_open()
+        self.logger.info("Stub Waveplate Home")
+
+        if not self.__stub_check_channel(1):
+            # Do nothing if channel is not enabled
+            return
+
+        self.current_position = 0
+        # Simulated delay for homing
+        # TODO: Delay calculation
+        time.sleep(1)
+        self.logger.info(f"Home Position: {self.current_position}")
```

Comment:
It occurred to me that, when logging, it's better to use [lazy evaluation](https://realpython.com/python-f-strings/#lazy-evaluation-in-logging) (and I hope that there are static analyzers for Python like there are for Java that detect this sort of thing).